oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/pipelined/PipelinedStrategy.java

```diff
@@ -0,0 +1,433 @@
+            ArrayBlockingQueue<NodeStateEntryBatch> emptyBatchesQueue = new ArrayBlockingQueue<>(numberOfBuffers);
+            // Queue with buffers filled by the transform task, used by the sort and save task. +1 for the SENTINEL
+            ArrayBlockingQueue<NodeStateEntryBatch> nonEmptyBatchesQueue = new ArrayBlockingQueue<>(numberOfBuffers + 1);
+
+            // Queue between sort-and-save thread and the merge-sorted-files thread
+            ArrayBlockingQueue<File> sortedFilesQueue = new ArrayBlockingQueue<>(64);
+
+            TransformStageStatistics transformStageStatistics = new TransformStageStatistics();
+
+            ScheduledFuture<?> monitorFuture = monitorThreadPool.scheduleWithFixedDelay(
+                    new MonitorTask(mongoDocQueue, emptyBatchesQueue, nonEmptyBatchesQueue, sortedFilesQueue, transformStageStatistics),
+                    10, 30, TimeUnit.SECONDS);
+
+            // Create empty buffers
+            for (int i = 0; i < numberOfBuffers; i++) {
+                emptyBatchesQueue.add(NodeStateEntryBatch.createNodeStateEntryBatch(bufferSizeBytes, maxNumberOfEntriesPerBuffer));
+            }
+
+            Stopwatch start = Stopwatch.createStarted();
+            MongoCollection<BasicDBObject> dbCollection = MongoDocumentStoreHelper.getDBCollection(docStore, Collection.NODES);
+            PipelinedMongoDownloadTask downloadTask = new PipelinedMongoDownloadTask(dbCollection, mongoBatchSize, mongoDocQueue);
+            ecs.submit(downloadTask);
+
+            File flatFileStore = null;
+
+            for (int i = 0; i < transformThreads; i++) {
+                NodeStateEntryWriter entryWriter = new NodeStateEntryWriter(blobStore);
+                PipelinedTransformTask transformTask = new PipelinedTransformTask(
+                        docStore,
+                        documentNodeStore,
+                        Collection.NODES,
+                        rootRevision,
+                        pathPredicate,
+                        entryWriter,
+                        mongoDocQueue,
+                        emptyBatchesQueue,
+                        nonEmptyBatchesQueue,
+                        transformStageStatistics
+                );
+                ecs.submit(transformTask);
+            }
+
+            PipelinedSortBatchTask sortTask = new PipelinedSortBatchTask(
+                    storeDir, pathComparator, algorithm, emptyBatchesQueue, nonEmptyBatchesQueue, sortedFilesQueue
+            );
+            ecs.submit(sortTask);
+
+            PipelinedMergeSortTask mergeSortTask = new PipelinedMergeSortTask(storeDir, pathComparator, algorithm, sortedFilesQueue);
+            ecs.submit(mergeSortTask);
+
+
+            try {
+                LOG.info("Waiting for tasks to complete.");
+                int tasksFinished = 0;
+                int transformTasksFinished = 0;
+                while (tasksFinished < numberOfThreads) {
```

Comment:
Would it be possible to refactor some of the code in this function into more separate functions? The function itself is quite long and with 3 nested `try-catch blocks`. While this might be necessary, I think it would at least improve the readability for the future. I asked ChatGPT to do it for an older version of the code, and the result was quite nice. 

For example: 
```java
private void waitForTasksToComplete(int numberOfThreads, int transformThreads, int entryCount, Stopwatch start,
        ArrayBlockingQueue<BasicDBObject[]> mongoDocQueue,
        ArrayBlockingQueue<NodeStateEntryBatch> emptyBatchesQueue,
        ArrayBlockingQueue<NodeStateEntryBatch> nonEmptyBatchesQueue,
        ArrayBlockingQueue<File> sortedFilesQueue,
        TransformStageStatistics transformStageStatistics) {
        int tasksFinished = 0;
        int transformTasksFinished = 0;
        while (tasksFinished < numberOfThreads) {
            Future<?> completedTask = ecs.take();
         ...
```

And then we could call it:

```java
try {
    waitForTasksToComplete(ecs, numberOfThreads, monitorFuture, mongoDocQueue, emptyBatchesQueue,
        nonEmptyBatchesQueue, sortedFilesQueue, transformStageStatistics, start);
    flatFileStore = mergeSortedFilesTask.getFlatFileStoreFile();
    return flatFileStore;
} finally {
    shutdownExecutors();
}
```

Reply:
I agree that the method is very long and somewhat hard to read. But the refactoring proposed above is also not very satisfying, as the method has too many arguments. I'll keep this in mind for a future revision of this class, hopefully we can come up with a cleaner and more logical refactor. 

For the final changes to this PR, I'd rather focus only on correctness or on small changes, and avoid doing any refactoring other than trivial ones.